src/java/org/apache/cassandra/locator/AzureSnitch.java

```diff
@@ -0,0 +1,190 @@
+        static final String METADATA_HEADER = "Metadata";
+        static final String API_VERSION_PROPERTY_KEY = "azure_api_version";
+
+        private final String apiVersion;
+
+        protected AzureCloudConnector(SnitchProperties properties) throws IOException
+        {
+            super(properties.get(METADATA_SERVICE_PROPERTY_KEY, DEFAULT_METADATA_SERVICE_URL));
+            apiVersion = getApiVersion(properties);
+            logger.debug("Resolved API version " + apiVersion);
+        }
+
+        String getResponse() throws IOException
+        {
+            return apiCall(metadataServiceUrl,
+                           String.format(METADATA_QUERY, apiVersion),
+                           "GET",
+                           ImmutableMap.of(METADATA_HEADER, "true"),
+                           200);
+        }
+
+        String getApiVersion(SnitchProperties properties) throws IOException
+        {
+            String version = properties.get(API_VERSION_PROPERTY_KEY, null);
+
+            if (version != null && API_VERSION_PATTERN.matcher(version).matches())
+                return version;
+
+            logger.debug(format("API version of '%s' does not follow pattern %s or is null. Going to " +
+                                "parse the latest version from the service ...",
+                                version, API_VERSION_PATTERN));
+
+            String response = apiCall(metadataServiceUrl,
+                                      VERSIONS_ENDPOINT,
+                                      "GET",
+                                      ImmutableMap.of("Metadata", "true"),
+                                      200);
+
+            JsonNode versions = JsonUtils.JSON_OBJECT_MAPPER.readTree(response);
+            JsonNode apiVersions = versions.get("apiVersions");
+
+            if (apiVersions == null)
+                throw new IllegalStateException("Unable to get field 'apiVersions' from the response!");
+
+            if (!(apiVersions.isArray()))
+                throw new IOException(format(
+                "Expected a list, but got a %s:%s", apiVersions.getClass().getSimpleName(), apiVersions));
+
+            String[] versionsArray = JsonUtils.JSON_OBJECT_MAPPER.convertValue(apiVersions, String[].class);
+
+            if (versionsArray != null && versionsArray.length > 0)
+                return versionsArray[versionsArray.length - 1];
+            else
+                throw new IllegalStateException("returned apiVersions array is empty!");
+        }
+
+        @Override
+        public String toString()
+        {
+            return format("%s{%s=%s,%s=%s}",
+                          AzureCloudConnector.class.getName(),
+                          METADATA_SERVICE_PROPERTY_KEY,
+                          metadataServiceUrl,
+                          API_VERSION_PROPERTY_KEY,
+                          apiVersion);
+        }
+    }
+
+    public String getRack(InetAddressAndPort endpoint)
```

Comment:
This method, and `getDatacenter` method, are identical in all snitch implementations. The only difference is that it returns a value of a field `rack` and, say in EC2, it returns value of a field `ec2zone`. Could you extract simple method like `getRackInternal() { return rack }`, define it abstract in `AbstractNetworkTopologySnitch` and move the implementation of `getRack(endpoint)` to `AbstractNetworkTopologySnitch`? You can do the same for `getDatacenter(endpoint)` method, and do that for all the snitch implementations.